### PR TITLE
Improve `Data.List.Base` (fix #2359; deprecate use of with #2123)

### LINKED ISSUE #2359: List of sub-optimal definitions in `Data.List.Base`
While looking at PR #2355 from @mildsunrise I noticed a ridiculous number of uses of `with`. These turn out to not be localized to `Data.List.Properties` but come from uses of `with` in `Data.List.Base`. The vast majority of them can be simplified, with a really nice knock-on effect to simplify quite a few proofs.

But I'm guessing this is yet another case of breaking change, so has to wait until 3.0? I think we should outlaw uses of `with` in function definitions, if favour of defining explicit 'helper' functions.

---

- [x] mapMaybe #2361 
- [ ] unalignWith
- [x] unfold #2364 
- [x] scanr -- but already in another PR? #2395 
- [x] initLast -- ditto?
- [x] unsnoc
- [x] takeWhile #2365 
- [x] dropWhile #2365
- [x] filter #2365
- [x] partition #2365
- [x] span #2365
- [x] linesBy #2366 
- [x] wordsBy #2366
- [x] derun #2365

### PR DESCRIPTION
**Pull request by jamesmckinna**
Refactor towards `if_then_else_`:
* `wordsBy`
* `linesBy`

### CODE REVIEW

**Review by JacquesCarette — APPROVED** [review]

File: src/Data/List/Base.agda [hunk 1]
```diff
@@ -428,9 +428,10 @@ linesBy {A = A} P? = go nothing where
 
   go : Maybe (List A) → List A → List (List A)
   go acc []       = maybe′ ([_] ∘′ reverse) [] acc
-  go acc (c ∷ cs) with does (P? c)
-  ... | true  = reverse (Maybe.fromMaybe [] acc) ∷ go nothing cs
-  ... | false = go (just (c ∷ Maybe.fromMaybe [] acc)) cs
+  go acc (c ∷ cs) = let acc′ = Maybe.fromMaybe [] acc in
```
Comment (MatthewDaggitt) [inline src/Data/List/Base.agda:431]: Any chance we could lay this out as:
```
go acc (c ∷ cs) =
  let acc′ = Maybe.fromMaybe [] acc in
  if does (P? c)
    then reverse acc′ ∷ go nothing cs
    else go (just (c ∷ acc′)) cs
```

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Otherwise looks good!

File: src/Data/List/Base.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Data/List/Base.agda:431]: Will do (eventually)

Comment (jamesmckinna) [reply src/Data/List/Base.agda]: Done.

**Review by Taneb — APPROVED** [review]

File: src/Data/List/Base.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Data/List/Base.agda]: Apologies, I realise I didn't do *precisely* what you asked for, and then re-read #2369 so have updated to use a `where` clause now... and now the layout is consistent with `wordsBy` in L452-L454 below...

**Review by MatthewDaggitt — APPROVED** [review]

**Pull request merged**